### PR TITLE
[RNMobile] Add featured image settings to the Latest Posts block

### LINKED ISSUE #4034: Feature Request: Add support to featured image settings to the Latest Posts block
**Is your feature request related to a problem? Please describe.**
It was requested in 4326230-zd-woothemes to add support to featured image settings for the Latest Posts block.

**Describe the solution you'd like**
Add the support of "Featured image settings" to the Latest Posts block settings in the app according to Calypso's options: https://wordpress.com/support/wordpress-editor/blocks/latest-posts-block/#featured-image-settings

### LINKED ISSUE #29815: Mobile: Latest Posts Block: Review Missing Settings Between Web and App
## Description

The latest posts block is currently missing some settings from the apps that are available on the web. Specifically, the following options aren't available to users from the mobile apps:

**Post meta settings**

- [ ] Display author name.

**Featured image settings** [Reported in wordpress-mobile/gutenberg-mobile/issues/4034]

- [x] Display featured image.
- [ ] Image size.
- [ ] Image dimensions.
- [x] Image alignment.
- [x] Add link to featured image.

**Sorting and filtering**

- [ ] Author.
- [ ] Categories. [Reported in wordpress-mobile/gutenberg-mobile/issues/2069]

It'd be great to review these differences and determine which settings should be made available in the apps.

## Screenshots

| Web settings |  |
| ------------- | ------------- |
|  |  |

| App settings |  |
| ------------- | ------------- |
|  |  |

## WordPress information
- WordPress version: 5.7
- Gutenberg version: 10.1.1
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

## Device information
- Device: Pixel 5
- Operating system: Android 11
- WordPress app version: alpha-280

_Note, although I only tested on Android, the same settings are present for iOS._

### PR DESCRIPTION
**Pull request by SiobhyB**
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4034
Partially fixes https://github.com/WordPress/gutenberg/issues/29815

## Related PRs

* `gutenberg-mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4674

## Description

This PR adds featured image settings to the native Latest Posts block. This is a step towards bringing the available block settings on the native side in alignment with the web. If a user selects the new `Display featured image` toggle, the following two options for customizing the featured image's appearance will be unveiled: 

* Image alignment.
* Add link to featured image.

_Note, the option to specify the featured image's size/dimensions will be addressed separately to this PR. This is because these options are a bit more complex and may warrant some updates to the image block's settings._

_Please refer to the `Types of changes` section below for an overview of the code changes._

### Testing changes in the app

 _Prerequisite: A WordPress site with more than one published blog post, with a featured image assigned to at least one of the posts._

* Open a post or page in the WordPress mobile app and add the Latest Posts block.
* Tap the cog/gear icon to access the block's settings.
* Confirm the new **Featured image settings** section appear within the block's settings and toggle the options.
* Preview the post to confirm that toggling the new settings work as expected i.e. you should be able to view any featured images associated with the latest posts and their alignment should change if adjusted.

### Testing for any web regressions

As these PR introduces some code changes to the web's alignment controls, there is a chance for unintentional regressions. We should therefore test that the alignment controls continue to work as expected while checked out with this branch on the web. Places where the controls appear include within the settings for the Image and Latest Posts blocks.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
|  |  |

## Video

https://user-images.githubusercontent.com/2998162/157964542-0a0ca502-3900-4943-bc5b-778b81858356.mov

## Types of changes

New feature (non-breaking change which adds functionality), with the following high-level overview of the code:

* A [new `Featured image settings` section has been added](https://github.com/WordPress/gutenberg/blob/9278d5fdce0842c6032d82596f3311bd800e1100/packages/block-library/src/latest-posts/edit.native.js#L200) to the Latest Posts settings, following the same code pattern used for other sections.
* If a user selects [the new `Display featured image` toggle](https://github.com/WordPress/gutenberg/blob/9278d5fdce0842c6032d82596f3311bd800e1100/packages/block-library/src/latest-posts/edit.native.js#L201), the following two options for customizing the featured image's appearance will be unveiled: 
  * 1. Change the image's alignment. 
  * 2. Add a link to the image. 
* This PR re-uses [the `BlockAlignmentControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/alignment-control) to implement the alignment options, which required the following changes:
   * As a significant number of native-specific changes were required, [a native version of the component's ui.js file](https://github.com/WordPress/gutenberg/blob/abf44c8854d19d8726dc6601893ea39e564a884a/packages/block-editor/src/components/block-alignment-control/ui.native.js) was created. The main removal/additions to the original file's code can be found in https://github.com/WordPress/gutenberg/pull/39257/commits/ec59670ae3d7de5aa89c495733de87d418843887.
   * All native-specific parts of the code have been removed from the original ui.js, as they're now redundant. The specific changes can be found in https://github.com/WordPress/gutenberg/pull/39257/commits/40765b409fb2f4e7a401fc18f30e3306a1fa3ef4 and https://github.com/WordPress/gutenberg/pull/39257/commits/9fb7b3893038471152f171170c479df27107a22a.
   * [A new constants.js file](https://github.com/WordPress/gutenberg/blob/abf44c8854d19d8726dc6601893ea39e564a884a/packages/block-editor/src/components/block-alignment-control/constants.js) has been created to add constants that are shared by both the native and web files, such as `BLOCK_ALIGNMENTS_CONTROLS`.  
   * The changes that were made in the native version of `BlockAlignmentControl` enable [the `BottomSheetSelectControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/mobile/bottom-sheet-select-control) to be rendered when the `isBottomSheetControl` prop is set to `true`. This was a desired change, as it'll make it easier to add alignment options within a block's bottom sheet settings (rather than only the toolbar) in the future.
   * `BottomSheetSelectControl` was [updated to allow icons](https://github.com/WordPress/gutenberg/blob/21e52a1024810309ca95461c62596baa9f33d946/packages/components/src/mobile/bottom-sheet-select-control/index.native.js#L85) to be displayed for each option in the subsheet, enabling the use of the alignment icons. It's likely that the use of icons will be a future use case for other developers using this component, too.
* _Note, the option to specify the image's size/dimensions will be addressed separately to this PR._ 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

**Review by fluiddot — COMMENTED** [review]
@SiobhyB thanks for adding the featured image settings to the latest posts block, great work 🎊 .

I tested locally and the changes work as expected, however, I'm wondering whether to use the dropdown for displaying the alignment options. My concern comes from the fact that for other settings like colors, this is done within a subscreen instead of displaying another bottom sheet over the main one. The approach you followed is ok from my side, but in order to keep consistency with the rest of elements within block settings, we might consider that option, wdyt?

I understand that the change I'm suggesting implies applying several changes to the PR, so please don't take it as a mandatory change but just as a suggestion. I'll be more than happy to discuss the best approach and if eventually, we consider continuing with this PR I'll be more than happy to approve it. I'd like also to have a second opinion from another reviewer regarding using a subscreen vs bottom sheet, @jhnstn what's your opinion?

**Comment by SiobhyB** [comment]
@fluiddot, thank you for your review! I think you're right. 

I was pretty focused on re-using [the `BlockAlignmentControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/alignment-control) but, upon reflection, I don't think re-using brings many benefits in this case. In fact, it's created some complexity and, as you've pointed out, some inconsistency.  

Edited to add: I've gone ahead to refactor using [the `BottomSheetSelectControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet-select-control/README.md) instead. I saw that component uses `BottomSheet.SubSheet`, which is used in other places, like for font sizes. Does that refactor seem more inline with what you'd expect? Looking forward to hearing your feedback.

**Comment by fluiddot** [comment]
> @fluiddot, thank you for your review! I think you're right.
> 
> I was pretty focused on re-using [the `BlockAlignmentControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/alignment-control) but, upon reflection, I don't think re-using brings many benefits in this case. In fact, it's created some complexity and, as you've pointed out, some inconsistency.

Makes sense to me, I think the implementation was right but my main concern was about consistency.

> Edited to add: I've gone ahead to refactor using [the `BottomSheetSelectControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet-select-control/README.md) instead. I saw that component uses `BottomSheet.SubSheet`, which is used in other places, like for font sizes. Does that refactor seem more inline with what you'd expect? Looking forward to hearing your feedback.

Great ⭐, thanks @SiobhyB for updating the PR 🙇 ! I like the idea of using the `BottomSheetSelectControl` component, as it looks like we could provide the same functionality with that component. I'll try to spare some time to review this but after a quick review, I think the implementation looks solid. I'll post a proper PR review once I jump into a deeper look.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (fluiddot) [inline packages/block-library/src/latest-posts/edit.native.js]: In the future, it would be great if `BlockAlignmentControl` renders this component (i.e. `BottomSheetSelectControl`) underneath, this way we won't need to set up the options and we could use it here.

I was checking [the previous commits](https://github.com/WordPress/gutenberg/pull/39257/commits/9d99ffa0d6850e4127b8e202ead005aaa574a58b) where you introduced `isBottomSheetControl` prop in `BlockAlignmentUI`, and I was wondering if we could follow that approach for this. But, I noticed that it most likely imply a big refactor because we would need to implement the native version of the Block Alignment Toolbar component to provide [here](https://github.com/WordPress/gutenberg/blob/fcd2e769654e272c270f7f725e4087cbe0f3ed8d/packages/block-editor/src/components/block-alignment-control/ui.js#L86) the `BottomSheetSelectControl` component.

File: packages/block-library/src/latest-posts/constants.js
Comment (fluiddot) [inline packages/block-library/src/latest-posts/constants.js]: Not sure if we should expose a native-only constant in the constants file, if it's going to be used only in the `edit.native.js` file I might consider defining the constant there, wdyt?

**Review by fluiddot — COMMENTED** [review]
I performed another test on Android and iOS and the changes worked as expected ✅. I only added a comment regarding where to place the `BLOCK_ALIGNMENTS_CONTROLS` constant, apart from that the PR looks great to me 🎊 .

**Comment by fluiddot** [comment]
@SiobhyB I also noticed that the PR checks look frozen, maybe updating with `trunk` branch re-triggers them.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (SiobhyB) [reply packages/block-library/src/latest-posts/edit.native.js]: @fluiddot, thank you for the suggestion! I've gone ahead with the proposed refactor, introducing the `isBottomSheetControl` prop again. As this required a significant number of changes to the `ui.js` file, I created a separate native version in https://github.com/WordPress/gutenberg/pull/39257/commits/19b0e215109e85af346c74bffe2e8de862439626.

File: packages/block-library/src/latest-posts/constants.js
Comment (SiobhyB) [reply packages/block-library/src/latest-posts/constants.js]: The changes to this shared constants file have been removed following the recent refactor. 👍

**Comment by SiobhyB** [comment]
@fluiddot, thank you for your suggestions! I've gone ahead with a refactor, so that `BlockAlignmentControl` is now able to renders `BottomSheetSelectControl` underneath. Looking forward to hearing what you think whenever you have the chance.

File: packages/block-library/src/latest-posts/edit.native.js
Comment (fluiddot) [reply packages/block-library/src/latest-posts/edit.native.js]: Great, thank you very much @SiobhyB for tackling this 🙇.

I'd like to clarify that with [my previous comment](https://github.com/WordPress/gutenberg/pull/39257#discussion_r837369643) I didn't want to imply that we had to do the refactor in this PR, but I'm happy to see you sorted it out 😊.  I'll try to take a look as soon as possible at the latest changes.

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! Awesome work @SiobhyB 🏅 !

I performed different tests on both native and web platforms and the changes work as expected.

_Tested on Samsung Galaxy S20 FE 5G (Android 11) and Simulator iPhone 13 (iOS 15.4)._

**Comment by SiobhyB** [comment]
Thanks so much @fluiddot! 🙇‍♀️

**Pull request merged**
